feat(undo): re-sync undo files on external change

yilisharcs opened the pull request:
Problem: External changes cause the hashes of a file and its undo file to de-sync, leading to loss of undo history.

Solution: On hash mismatch, overwrite the buffer with the backup file. Underlying implementation requires that every `:write` operation happen twice to keep the original and backup files in sync. Provides 'undosync' option, reliant on 'backup'.

fix #33666

yilisharcs committed:
feat(undo)!: re-sync undo files on external change

Problem: external changes cause the hashes of a file and its undo file
to de-sync, leading to loss of undo history.

Solution: keep a backup of the file to check against if the original has
been changed.

BREAKING CHANGE: modify "backup" to save a copy of the file _after_ it
is written instead of _before_.

Addresses #33666

yilisharcs commented:
Instead of manually reading and writing buffers, I opted to call the built-in readfile and buf_write functions to do the heavy lifting. This should be much more stable.

It doesn't crash if binary files are overwritten by plaintext anymore.

justinmk commented on src/nvim/undo.c:
```diff
@@ -3198,6 +3204,135 @@ void f_undotree(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+  char *dirp = p_bdir;
+  while (*dirp) {
+    // Isolate one directory name from the comma-separated list
+    size_t dir_len = copy_option_part(&dirp, IObuff, IOSIZE, ",");
```
surely this logic already exists elsewhere (where `p_bdir` is processed), can we re-use it instead of duplicating it

yilisharcs replied:
In `buf_write_make_backup`, but that's a big and scary function and it writes instead of reads. I didn't know where else to look, so I decided to duplicate it and find a better solution later

yilisharcs commented on runtime/doc/options.txt:
```diff
@@ -7064,6 +7065,19 @@ A jump table for the options with a short description can be found at |Q_op|.
+			global
+	Vim will detect if an undo file is no longer synchronized with the
+	file it was written for (with a hash of the file contents) and
+	ignore it when the file was changed after the undo file was written,
```
I copied this part verbatim from undo.txt at undo-persistance up to "If 'undosync'...". I myself don't understand the corruption issues as much as I should, but "ignore it" means to fallback to default behavior, which is discarding history if the hash doesn't match.

justinmk commented on runtime/doc/options.txt:
```diff
@@ -7064,6 +7065,19 @@ A jump table for the options with a short description can be found at |Q_op|.
+	file it was written for (with a hash of the file contents) and
+	ignore it when the file was changed after the undo file was written,
+	to prevent corruption. If 'undosync' and 'backup' are enabled, the
+	backup file at 'backupdir' can be used to safely re-synchronize the
```
Let's go ahead and plow through and enable this behavior as a new feature of 'backup'. We can mention it in news.txt.

If people object we can always revisit later.

yilisharcs replied:
Took a peek at the documentation and concluded that this is a daunting task. There's a lot of logic behind the scenes for creating backups involving patchmode, backupcopy, and writebackup. Even the name of the option would become misleading given that it operates on a different version of the file than both patchmode and writebackup. Imo this really is the option of least friction

yilisharcs replied:
Took a closer look at the code and concluded having this as a new feature of backup is as simple as running a copy operation after everything is said and done, no need to disentangle every other option from it. Ofc the documentation needs to be updated to match, but yeah. I'll make another commit to remove undosync

yilisharcs commented:
This can yet be improved. Instead of having write calls happen twice, we can have `buf_write_make_backup` be called again in `buf_write`. I'll look into this later today

yilisharcs commented:
Failing tests all concern changed behavior, because the backups aren't the same anymore due to the copy operation at the end of the function. How can I help with this?

zeertzjq commented:
> result in empty undo history when later open the file in vim.

This is not a behavior of backup. It's a behavior of undo.

justinmk commented:
Yes, but that doesn't change any of my comment

zeertzjq commented:
Yes, it does, because no one is arguing about preserving this behavior of undo, and undo and backup have been two independent features.

justinmk commented on src/nvim/bufwrite.c:
```diff
@@ -715,13 +715,47 @@ static int get_fileinfo(buf_T *buf, char *fname, bool overwriting, bool forceit,
 }
 
+/// @return The backup file name
+char *buf_get_backup_name(char *fname, char **dirp, bool no_prepend_dot)
```
this is a nice refactor that could be merged pretty quickly if extracted to another PR

yilisharcs committed:
feat(undo): add check for large undo files

Problem: new undo syncing feature can cause undo files to grow rapidly
when switching between two branches of version control, which could
cause performance issues.

Solution:
- checkhealth module scans 'undodir' for files larger than 10MB
- warn users once on write that their undo file is too large

justinmk commented:
If you plan to continue this please keep it rebased. When someone has the time to think carefully about it, then it will be ready to merge.

Even though there's an open question, I don't think that blocks test coverage? Regardless of how to decide to enable this feature, if that part changes it's only a small change to the tests.

One other obvious blocker is that the docs just aren't clear. How will users understand how to use this and whether they should use it? Ideally it should "just work". That is what I was [driving at](https://github.com/neovim/neovim/pull/36262#discussion_r2449090124), though maybe it's blocked for a technical reason I don't understand.

yilisharcs committed:
docs: new "backup" behavior

yilisharcs commented:
aaand we're done! the docs should be clearer now, i hope. also "backup" defaults to true now

yilisharcs committed:
fix(undo): abort buffer replace if file is readonly
